testing: include region tags in XML test output

tbpg opened the pull request:
This can be used to associate a region tag with the test that covers it.

Current coverage:
```
573/1126 (50.89%) of region tags are tested.
```

See the package comment for additional details.

This ends up with XML like:

```xml
<testsuites>
	<testsuite tests="12" failures="0" time="0.039" name="github.com/GoogleCloudPlatform/golang-samples/functions/http">
		<properties>
			<property name="go.version" value="go1.14"></property>
		</properties>
		<testcase classname="http" name="TestCORSEnabledFunctionAuth" time="0.000">
			<properties>
				<property name="region_tags" value="functions_http_cors_auth"></property>
			</properties>
		</testcase>
		<testcase classname="http" name="TestCORSEnabledFunctionAuthPOST" time="0.000">
			<properties>
				<property name="region_tags" value="functions_http_cors_auth"></property>
			</properties>
		</testcase>
```

BenWhitehead commented on testing/sampletests/main.go:
```diff
@@ -0,0 +1,245 @@
+// The test coverage over all regions is printed at the end.
+//
+// To get the number of unique region tags in the repo manually, run:
+//     grep -RoPh '\[START \K(.+)\]' | sort | uniq | wc -l
```
you could also use `sort -u` in place of `sort | uniq` so that it's all done when sorting. For large datasets this tends to be faster.

tbpg replied:
Done.

ace-n commented on testing/sampletests/main.go:
```diff
@@ -0,0 +1,245 @@
+			if !exact {
+				return nil, fmt.Errorf("PathEnclosingInterval got not exact path for %q in %v", id.Name, file)
+			}
+			ast.Inspect(path[1], func(node ast.Node) bool {
```
I'm assuming this `ast` utility won't translate to other languages?

tbpg replied:
No, it's specific to Go. Other languages might have other AST parsers, but this one won't work.

broady commented on testing/sampletests/main.go:
```diff
@@ -0,0 +1,245 @@
+)
+
+var startRe = regexp.MustCompile("\\[START ([^\\]]+)\\]")
+var endRe = regexp.MustCompile("\\[END ([^\\]]+)\\]")
```
```suggestion
var (
	startRe = regexp.MustCompile("\\[START ([^\\]]+)\\]")
	endRe   = regexp.MustCompile("\\[END ([^\\]]+)\\]")
)
```

tbpg replied:
Done.

broady commented on testing/sampletests/main.go:
```diff
@@ -0,0 +1,245 @@
+// name. It reads from stdin and writes to stdout.
+//
+// For example:
+//     TestFoo -> TestFoo__foo_hello_world__foo_hello_gopher
```
in this case, are there two regions?

tbpg replied:
Yes. Updated the comment.

broady commented on testing/sampletests/main.go:
```diff
@@ -0,0 +1,245 @@
+				if !ok {
+					return true
+				}
+				calleeObj := typeutil.Callee(pkg.TypesInfo, call)
```
will this get callees of function references?

e.g.

```
func Foo() {}
func Bar() {}
func Baz() {}

func TestX() {
	var funcs []func() = []func(){Foo, Bar, Baz}
	...
}
```

tbpg replied:
No, I don't think so. The `CallExpr` where those functions are called will end up being in a `for` loop over `funcs`. Would need to look for transitive function calls. `TestX` uses/references `Foo`, which calls `DoFoo`, which is part of region `do_foo`.

broady commented:
An idea:
https://godoc.org/golang.org/x/tools/go/callgraph

Another idea:
Is there call graph information in the test coverage output? I don't think so but that would seem promising

ace-n commented:
@broady we've already considered coverage output, and found that it didn't really suit our needs. (I'll ping you a link internally to the design doc section that covers this.)

tbpg commented:
> https://godoc.org/golang.org/x/tools/go/callgraph
Looks promising... Looking.

> Is there call graph information in the test coverage output? I don't think so but that would seem promising
Unfortunately, no. Coverage only says if a line ran, not what test caused it to run (Java does). We could run coverage per-test, but that would be too slow.

broady commented:
> Unfortunately, no. Coverage only says if a line ran, not what test caused it to run (Java does). We could run coverage per-test, but that would be too slow.

+1, and some tests would need to be re-written. some packages depend on test ordering:

e.g.
```
func TestCreateBucket(...) {}
func TestCreateObject(...) {}
```

tbpg commented:
Pending approval from @ace-n, @broady, and Kokoro, I think this is ready to go.

I changed the behavior of this to add `properties` to the test cases rather than change the names of the tests. I also changed the logic to actually look within each package, rather than doing a global search and replace, so the results are more accurate.

@broady, I tried using `callgraph` (as you can see from the commit history), but I couldn't get it working the same way. I'm sure it's possible, but I think this works for today.

ace-n commented:
LGTM assuming you've confirmed our internal systems can access this.

(They'll have to be updated accordingly; so maybe that's something we should sync w/ @jmdobry at some point about.)

codyoss commented on testing/sampletests/main.go:
```diff
@@ -0,0 +1,260 @@
+				region := regions[0][1]
+				uniqueRegionTags[region] = struct{}{}
+				for _, r := range testFileRanges[path] {
+					if r.start >= lineNum && r.end >= lineNum {
```
Should these both be `>=`?

codyoss commented on testing/sampletests/main.go:
```diff
@@ -0,0 +1,365 @@
+				if callee == nil {
+					return true
+				}
+				if callee.Pkg() != obj.Pkg() {
```
Do we ever test from external packages in this repo? "hello" != "hello_test". Should this be a special case?

tbpg replied:
Good idea! I checked and there are some `_test` packages, so I added support.

codyoss commented:
LGTM, just the one final questions. Having the tests and being able to play with them made this much easier to reason about. Thanks for adding those!
